### PR TITLE
Feature/glblagctwo 1094

### PR DESCRIPTION
**Pull request by broguinn**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
+var hl = require('highland');
+var Promise = require('bluebird');
+var redis = require('then-redis');
+var retry = require('retry');
```
Comment (mavdi) [inline lib/index.js:4]: ooooooo this looks like an interesting lib

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        return Promise.resolve([undefined, constructor]);
+    }
+
+    function createHighlandStream(id, streamConstructor) {
+        var tryStream;
+        try {
+            tryStream = streamConstructor(id);
+        } catch(err) {
+            throw new Error("Could not establish SSE stream", err);
```
Comment (mavdi) [inline lib/index.js]: Are we capturing this error somewhere? Seems it will crash the app.

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    function tryCallback(err, chunk, push, next) {
+        if (!consumer.eventHook) {
+            push(err, chunk);
+            return next();
+        }
+
+        var operation = retry.operation({
+            retries: 10000,
```
Comment (kristofsajdak) [inline lib/index.js]: Don't mind that we have the option of setting a max amount of attempts, however the default should be to retry forever given how we apply the SSE pattern overall with queuing

Setting the forever option is possible it seems : https://github.com/tim-kos/node-retry/blob/d0b863bb87b486a77eec20d6ec3f5fd2f35c6771/test/integration/test-retry-operation.js#L84

We could add in another ticket to change the signature a bit allowing optional overrides?

``` js
var consumeRetryOptions = {
  retries: 5,
  factor: 3,
  minTimeout: 1 * 1000,
  maxTimeout: 60 * 1000,
  randomize: true,
}
consumer(consumeRetryOptions)
    .consume(function makeSSEStream(lastEventId) {
        return grabToken().then(function(token) {
            return request({
                uri: 'http://myapi.example.com/changes/stream?resources=trackingData,equipment',
                headers: {'Authorization': 'Bearer ' + token, 'last-event-id': lastEventId}
            });
        })
    })
    .onEvent(function (sse) {
        return new Promise(function (resolve) {
            exchange.publish({text: sse.data}, {key: sse.event}).on('drain', resolve);
        })
    })
    .checkpoint({redisUrl: 'redis://someuser:secret@localhost:10242', messages: 5});
```

default would be

``` js
consumer()
    .consume(function makeSSEStream(lastEventId) {
// ...

```

File: lib/index.js (hunk 1 above)
Comment (kristofsajdak) [inline lib/index.js:4]: Agree, great find!

File: test/sse.spec.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        it('creates a checkpoint after receiving 1 message', function() {
+            consumer
+                .consume(createStream(streamThreeURL))
+                .checkpoint({
+                    redisUrl: redisUrl,
+                    messages: 1
+                });
+
+            return Promise.delay(50).then(function(){
+                return redisClient.get('checkpoint')
```
Comment (kristofsajdak) [inline test/sse.spec.js:137]: Let's think of something else than time to drive off assertions, it's not very deterministic 

How about an optional callback parameter in .checkpoint? 

``` js
 consumer
            .consume(createStream(streamThreeURL))
            .checkpoint({
                    redisUrl: redisUrl,
                    messages: 1
                }, 
                // callback is invoked whenever a checkpoint is set 
                function (checkpoint, chunk) {
                     //... perform assertions here
                    // both checkpoint and the event chunk are available 
                    done()
                });
```

Comment (kristofsajdak) [inline test/sse.spec.js:137]: A Promise based alternative could work as well

File: lib/index.js (hunk 3 above)
Comment (broguinn) [inline lib/index.js]: Resolved.

File: test/sse.spec.js (hunk 4 above)
Comment (broguinn) [inline test/sse.spec.js:137]: Done

File: test/sse.spec.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+        var server = app.listen(port, function() {
+            var host = server.address().address;
+            var port = server.address().port;
+            console.log('Test server listening at http://%s:%s', host, port);
+        });
+    });
+
+    beforeEach(function() {
+        consumer = new Consumer(redisUrl);
```
Comment (kristofsajdak) [inline test/sse.spec.js]: redisUrl can be passed in either through the Consumer and checkpoint function 
Any reason we would want to pass it in through the former? The redis persistence seems like a checkpoint concern.

Comment (broguinn) [inline test/sse.spec.js]: The consume function needs to be able to pass the last checkpoint id to the streamConstructor, which it needs to retrieve from Redis.

**Pull request merged**